transport/transporters/s3/transporter/transporter.go

```diff
@@ -0,0 +1,351 @@
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/Nextdoor/pg-bifrost.git/marshaller"
+	"github.com/Nextdoor/pg-bifrost.git/shutdown"
+	"github.com/Nextdoor/pg-bifrost.git/stats"
+	"github.com/Nextdoor/pg-bifrost.git/transport"
+	"github.com/Nextdoor/pg-bifrost.git/transport/batch"
+	"github.com/Nextdoor/pg-bifrost.git/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/cenkalti/backoff"
+	"github.com/cevaris/ordered_map"
+	"github.com/sirupsen/logrus"
+)
+
+// Generate mock for s3 client
+//go:generate mockgen -destination=mocks/mock_s3.go -package=mocks github.com/aws/aws-sdk-go/service/s3/s3iface S3API
+
+var (
+	TimeSource utils.TimeSource = utils.RealTime{}
+)
+
+// key_join is a helper to concatenate strings to form an S3 key
+func key_join(gzipped bool, strs ...string, ) string {
+	var sb strings.Builder
+	for i, str := range strs {
+		if str == "" || str == "/" {
+			continue
+		}
+
+		// clean any leading and trailing slashes
+		str = strings.TrimRight(str, "/")
+		str = strings.TrimLeft(str, "/")
+
+		// concatenate string and add trailing slash
+		sb.WriteString(str)
+
+		if i != len(strs) - 1 {
+			sb.WriteString("/")
+		}
+	}
+
+	if gzipped {
+		sb.WriteString(".gz")
+	}
+
+	return sb.String()
+}
+
+type S3Transporter struct {
+	shutdownHandler shutdown.ShutdownHandler
+
+	inputChan   <-chan transport.Batch         // receive a batch (slice) of MarshalledMessages
+	txnsWritten chan<- *ordered_map.OrderedMap // Map of <transaction:progress.Written>
+
+	statsChan chan stats.Stat
+
+	log         logrus.Entry
+	client      s3iface.S3API
+	bucketName  string
+	keySpace    string
+	retryPolicy backoff.BackOff
+}
+
+func NewTransporterWithInterface(shutdownHandler shutdown.ShutdownHandler,
+	inputChan <-chan transport.Batch,
+	txnsWritten chan<- *ordered_map.OrderedMap,
+	statsChan chan stats.Stat,
+	log logrus.Entry,
+	id int,
+	bucketName string,
+	keySpace string,
+	client s3iface.S3API,
+	retryPolicy backoff.BackOff) transport.Transporter {
+
+	log = *log.WithField("routine", "transporter").WithField("id", id)
+
+	return &S3Transporter{
+		shutdownHandler,
+		inputChan,
+		txnsWritten,
+		statsChan,
+		log,
+		client,
+		bucketName,
+		keySpace,
+		retryPolicy,
+	}
+}
+
+// NewTransporter returns a s3 transporter
+func NewTransporter(shutdownHandler shutdown.ShutdownHandler,
+	inputChan <-chan transport.Batch,
+	txnsWritten chan<- *ordered_map.OrderedMap,
+	statsChan chan stats.Stat,
+	log logrus.Entry,
+	id int,
+	bucketName string,
+	keySpace string,
+	retryPolicy backoff.BackOff,
+	awsRegion *string,
+	awsAccessKeyId *string,
+	awsSecretAccessKey *string,
+	endpoint *string,) transport.Transporter {
+
+	var sess *session.Session
+
+	creds := credentials.NewStaticCredentials(*awsAccessKeyId, *awsSecretAccessKey, "")
```

Comment:
If `awsAccessKeyId` and `awsSecretAccessKey` are not set then we should default to letting the aws go library figure it out from the env.